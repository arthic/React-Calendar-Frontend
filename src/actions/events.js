import Swal from "sweetalert2";
import { fetchConToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";
import { types } from "../types/types";

// Nuevo evento
export const eventStartAddNew = (event) => {
	return async(dispatch, getState) => {

		// Extraemos del state gracias a thunk
		const {uid, name} = getState().auth

		try {
			const resp = await fetchConToken('events', event, 'POST') // endpoint, payload, metodo
			const body = await resp.json()

			// Grabar en la BD
			if(body.ok) {
				event.id = body.evento.id
				event.user = {
					_id: uid,
					name: name
				}
				dispatch(eventAddNew(event))
			}
		} catch (error) {
			console.log(error);
		}
	}
}

const eventAddNew = (event) => ({
	type: types.eventAddNew,
	payload: event
})

// Evento Activo
export const eventSetActive = (event) => ({
	type: types.eventSetActive,
	payload: event
})
// Limpiar evento activo
export const eventClearActiveEvent = () => ({type: types.eventClearActiveEvent})

// Actualziar evento
export const eventStartUpdated = (event) => {
	return async(dispatch) => {

		try {
			console.log(event)
			const resp = await fetchConToken(`events/${event.id}`, event, 'PUT')
			const body = await resp.json()

			if(body.ok) {
				dispatch(eventUpdated(event))
			} else {
				console.log(body.msg)
				Swal.fire('Error', body.msg, 'error')
			}

		} catch (error) {
			console.log(error);
		}
	}
}

const eventUpdated = (event) => ({
	type: types.eventUpdated,
	payload: event
})

// Borrar evento
export const eventStartDelete = () => {
	return async(dispatch, getState) => {

		// Del store
		const {id} = getState().calendar.activeEvent

		try {
			const resp = await fetchConToken(`events/${id}`, {}, 'DELETE')
			const body = await resp.json()

			if(body.ok) {
				dispatch(eventDeleted())
			} else {
				console.log(body.msg)
				Swal.fire('Error', body.msg, 'error')
			}

		} catch (error) {
			console.log(error);
		}
	}
}
const eventDeleted = () => ({type: types.eventDeleted})

// Cargar eventos
export const eventStartLoading = () => {
	return async(dispatch) => {

		try {
			const resp = await fetchConToken('events')
			const body = await resp.json()

			// Preparar eventos compatibles con frontend "fechas"
			const events = prepareEvents(body.eventos)

			// Establece los eventos en el store
			dispatch(eventLoaded(events))
		} catch (error) {
			console.log(error)
		}
	}
}
const eventLoaded = (events) =>  ({
	type: types.eventLoaded,
	payload: events
})

// Va al actions/auth
export const eventLogout = () => ({type: types.eventLogout})
