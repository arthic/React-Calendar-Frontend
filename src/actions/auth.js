import { fetchConToken, fetchSinToken } from '../helpers/fetch'
import { types } from '../types/types'
import Swal from 'sweetalert2'
import { eventLogout } from './events'


export const startLogin = (email, password) => {

	// Gracias a redux-thunk disponemos de dispatch
	return async(dispatch) => {
		// Llamado al endpoint auth, osea el url
		const resp = await fetchSinToken('auth', {email, password}, 'POST')
		const body = await resp.json()

		// ok viene de la personalzacion del backend
		if(body.ok) {
			localStorage.setItem('token', body.token)
			// El token dura 2h, grabamos la fecha exacta en que se creo
			localStorage.setItem('token-init-date', new Date().getTime())

			// En el state
				dispatch( login({
					uid: body.uid,
					name: body.name
				}))
		} else {
			Swal.fire('Error', body.msg, 'error')
		}
	}
}

const login = (user) => ({
	type: types.authLogin,
	payload: user
})

export const startRegister = (email, password, name) => {
	// Gracias a redux-thunk disponemos de dispatch
	return async(dispatch) => {
		// Llamado al endpoint auth, osea el url
		const resp = await fetchSinToken('auth/new', {email, password, name}, 'POST')
		const body = await resp.json()

		// ok viene de la personalzacion del backend
		if(body.ok) {
			localStorage.setItem('token', body.token)
			// El token dura 2h, grabamos la fecha exacta en que se creo
			localStorage.setItem('token-init-date', new Date().getTime())

			// En el state
			dispatch( login({
				uid: body.uid,
				name: body.name
			}))
		} else {
			Swal.fire('Error', body.msg, 'error')
		}
	}
}

export const startChecking = () => {
	return async(dispatch) => {
		// Llamado al endpoint auth, osea el url
		const resp = await fetchConToken('auth/renew') //Por defecto es GET, queda en blanco
		const body = await resp.json()

		// ok viene de la personalzacion del backend
		if(body.ok) {
			localStorage.setItem('token', body.token)
			// El token dura 2h, grabamos la fecha exacta en que se creo
			localStorage.setItem('token-init-date', new Date().getTime())

			// En el state
			// Control de Pantalla de carga Calendar Screen
			setTimeout(() => {
				dispatch( login({
					uid: body.uid,
					name: body.name
				}))
			}, 1800);
		} else {
			// Control de Pantalla de carga Login Screen
			setTimeout(() => {
				dispatch(checkingFinish())
			}, 1800);
		}
	}
}

const checkingFinish = () => ({type: types.authCheckingFinish})

export const startLogout = () => {
	return (dispatch) => {
		localStorage.clear()
		dispatch(eventLogout())
		dispatch(logout())
	}
}

const logout = () => ({type: types.authLogout})