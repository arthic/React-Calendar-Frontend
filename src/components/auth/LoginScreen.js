import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Swal from 'sweetalert2';
import { startLogin, startRegister } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';
import { LoadingScreen } from '../ui/LoadingScreen';
import './login.css';

export const LoginScreen = () => {
    // Despachar el state
    const dispatch = useDispatch()
    // Login
    const [formLoginValues, handleLoginInputChange] = useForm({
        lEmail: 'usuario@muestra.com',
        lPassword: 'testing'
    })
    const {lEmail, lPassword} = formLoginValues

    const handleLogin = (e) => {
        e.preventDefault()
        dispatch(startLogin(lEmail, lPassword))
    }

    // Registro
    const [formRegisterValues, handleRegisterInputChange] = useForm({
        rName: '',
        rEmail: '',
        rPassword1: '',
        rPassword2: ''
    })
    const {rName, rEmail, rPassword1, rPassword2} = formRegisterValues

    const handleRegister = (e) => {
        e.preventDefault()

        // Validar contraseñas
        if(rPassword1 !== rPassword2) {
            return Swal.fire('Error', 'Las constraseñas deben de ser iguales', 'error')
        }
        if(rPassword1.length < 6) {
            return Swal.fire('Error', 'La constraseña debe ser mayor a 6 caracteres', 'error')
        }
        if(rName.length < 1) {
            return Swal.fire('Error', 'Nombre de usuario obligatorio', 'error')
        }

        dispatch(startRegister(rEmail, rPassword1, rName))
    }

    // Manejo Pantalla de carga
	const {checking} = useSelector(state => state.auth)
    // Si se esta checando = true
	if(checking) {
        return (
            <LoadingScreen />
		)
    }

    return (
        <div className="container login-container animate__animated animate__fadeIn">
            <div className="logo-content">
                <i className="far fa-calendar-alt"></i>
                <h1><strong>Calendar</strong> App</h1>
            </div>
            <div className="row animate__animated animate__fadeInUp">
                <div className="col-md-4 login-form-1">
                    <h3>Ingreso</h3>
                    <form
                        onSubmit={handleLogin}
                    >
                        <div className="form-group">
                            <input
                                type="text"
                                className="form-control"
                                placeholder="Correo"
                                name="lEmail"
                                value={lEmail}
                                onChange={handleLoginInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Contraseña"
                                name="lPassword"
                                value={lPassword}
                                onChange={handleLoginInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="submit"
                                className="btnSubmit"
                                value="Login"
                            />
                        </div>
                    </form>
                </div>

                <div className="col-md-4 login-form-2">
                    <h3>Registro</h3>
                    <form
                        onSubmit={handleRegister}
                    >
                        <div className="form-group">
                            <input
                                type="text"
                                className="form-control"
                                placeholder="Nombre"
                                name="rName"
                                value={rName}
                                onChange={handleRegisterInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="email"
                                className="form-control"
                                placeholder="Correo"
                                name="rEmail"
                                value={rEmail}
                                onChange={handleRegisterInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Contraseña"
                                name="rPassword1"
                                value={rPassword1}
                                onChange={handleRegisterInputChange}
                            />
                        </div>

                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Repita la contraseña"
                                name="rPassword2"
                                value={rPassword2}
                                onChange={handleRegisterInputChange}
                            />
                        </div>

                        <div className="form-group">
                            <input
                                type="submit"
                                className="btnSubmit"
                                value="Crear cuenta" />
                        </div>
                    </form>
                </div>
            </div>
        </div>
    )
}