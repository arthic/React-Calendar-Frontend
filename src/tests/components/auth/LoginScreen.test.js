import '@testing-library/jest-dom'
import React from 'react';
import {mount} from 'enzyme'
import {Provider} from 'react-redux'
import { LoginScreen } from '../../../components/auth/LoginScreen';
import { startLogin, startRegister } from '../../../actions/auth';
import Swal from 'sweetalert2';

// Configurar el store de redux | npm install redux-mock-store --save-dev
import configureStore from 'redux-mock-store'
// Thunk para que ejecute las func, de las func asincronas
import thunk from 'redux-thunk'

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = {
	auth: {
		checking: false, // Ya cargo
		uid: null // No autenticado
	}
}
const store = mockStore(initState)
// Mock del store
store.dispatch = jest.fn()

// Mock al startLogin
jest.mock('../../../actions/auth', () => ({
	startLogin: jest.fn(),
	startRegister: jest.fn()
}))
// Mock del Sweet Alert
jest.mock('sweetalert2', () => ({
	fire: jest.fn()
}))

const wrapper = mount(
	<Provider store={store}>
		<LoginScreen />
	</Provider>
)

describe('Testing en <LoginScreen />', () => {
	beforeEach(() => {
		jest.clearAllMocks()
	})

	test('Debe desplegarse correctamente', () => {
		expect(wrapper).toMatchSnapshot()
	})

	test('Debe llamar el dispatch del login', () => {
		wrapper.find('input[name="lEmail"]').simulate('change', {
			target: {
				name: 'lEmail',
				value: 'test@testing.com'
			}
		})
		wrapper.find('input[name="lPassword"]').simulate('change', {
			target: {
				name: 'lPassword',
				value: 't3st1ng'
			}
		})

		// at(0) encuentre la primera posición porque hay 2 form
		wrapper.find('form').at(0).prop('onSubmit')({
			preventDefault(){}
		})
		expect(startLogin).toHaveBeenCalledWith(
			'test@testing.com',
			't3st1ng'
		)
	})

	test('No hay registro si las contraseñas son diferentes', () => {
		wrapper.find('input[name="rPassword1"]').simulate('change', {
			target: {
				name: 'rPassword1',
				value: 't3st1ng'
			}
		})
		wrapper.find('input[name="rPassword2"]').simulate('change', {
			target: {
				name: 'rPassword2',
				value: 'testing'
			}
		})
		// 2° form (1)
		wrapper.find('form').at(1).prop('onSubmit')({
			preventDefault(){}
		})
		// No llamar a la accion startRegister
		expect(startRegister).not.toHaveBeenCalled()
		expect(Swal.fire).toHaveBeenCalledWith(
			'Error',
			'Las constraseñas deben de ser iguales',
			'error'
		)
	})

	test('Debe dispararse el registro con constraseñas iguales', () => {
		wrapper.find('input[name="rName"]').simulate('change', {
			target: {
				name: 'rName',
				value: 'tester'
			}
		})
		wrapper.find('input[name="rEmail"]').simulate('change', {
			target: {
				name: 'rEmail',
				value: 'tester@testing.com'
			}
		})
		wrapper.find('input[name="rPassword1"]').simulate('change', {
			target: {
				name: 'rPassword1',
				value: 't3st1ng'
			}
		})
		wrapper.find('input[name="rPassword2"]').simulate('change', {
			target: {
				name: 'rPassword2',
				value: 't3st1ng'
			}
		})
		// 2° form (1)
		wrapper.find('form').at(1).prop('onSubmit')({
			preventDefault(){}
		})

		// Invertidos
		expect(Swal.fire).not.toHaveBeenCalled()
		expect(startRegister).toHaveBeenCalledWith(
			'tester@testing.com',
			't3st1ng',
			'tester'
		)
	})
})


