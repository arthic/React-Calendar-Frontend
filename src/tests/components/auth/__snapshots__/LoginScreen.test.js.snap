// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing en <LoginScreen /> Debe desplegarse correctamente 1`] = `
<div
  className="container login-container animate__animated animate__fadeIn"
>
  <div
    className="logo-content"
  >
    <i
      className="far fa-calendar-alt"
    />
    <h1>
      <strong>
        Calendar
      </strong>
       App
    </h1>
  </div>
  <div
    className="row animate__animated animate__fadeInUp"
  >
    <div
      className="col-md-4 login-form-1"
    >
      <h3>
        Ingreso
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="text"
            value="usuario@muestra.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="testing"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-4 login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rName"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword1"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword2"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
